General numeric type support

graeme-a-stewart opened the pull request:
Support is added for general numerical types in this PR. `EEjet` and `PseudoJet` become parametrised on `T <: Real`. `ClusterSequence` stays parametrised on `J <: FourMomentum`, but is it really important to pass the *parametrised* jet collection into the constructor otherwise performance greatly suffers! (i.e., `ClusterSequence{EEjet{Float64}}` not `ClusterSequence{EEjet}`. To support this the `eltype` method is defined for both jet types to return the internal type used.

The `instrumented-jetreco.jl` script gets a new `--type` option that allows a numeric type to be passed for the reconstruction. Added a quick `benchmark.sh` script to run all three main use cases.

Current performance tests shows that we are losing a bit in the $pp$ `N2Tiled` reconstruction, so this needs to be fixed before merging. Performance for Durham $e^+e^-$ and $pp$ `N2Plain` are the same as the current v0.4.3 release.

Testing on M2pro gives

This branch:

```sh
 ~/.julia/dev/JetReconstruction/examples/ [general-numeric-types*] ./benchmark.sh 
pp 14TeV Tiled
Lowest time per event 189.185 μs
pp 14 TeV Plain
Lowest time per event 678.68667 μs
ee H Durham
Lowest time per event 28.38375 μs
```

v0.4.3:

```sh
 ~/.julia/dev/JetReconstruction/examples/ [general-numeric-types] ~/tmp/benchmark.sh
pp 14TeV Tiled
Lowest time per event 176.06666 μs
pp 14 TeV Plain
Lowest time per event 677.14125 μs
ee H Durham
Lowest time per event 27.92792 μs
```

Closes #91 

---
To do:
- [x] Convert `EEjet` to `T <: Real`
- [x] Convert `PseudoJet` to `T <: Real`
- [ ] Convert `TiledJet` to `T <: Real`
- [x] Add option to examples to switch types for the internal representation
  - [x] `instrumented_jetreco.jl`
  - [x] `jetreco.jl`
- [x] Ensure performance is maintained in `Float64` case
  - [x] pp Tiled
  - [x] pp Plain
  - [x] ee
- [ ] Add unit tests for `Float32` reconstruction

graeme-a-stewart committed:
Update PseudoJet to be type T <: Real

Allowing for different Float values, as well as more exotic types, if needed

graeme-a-stewart committed:
Update EEJet to be type T <: Real

Allowing for different Float values, as well as more exotic types,
if needed.

graeme-a-stewart committed:
Minor corrections to docstrings

graeme-a-stewart committed:
Fix types for EEJet

graeme-a-stewart committed:
Fix types for EE reco

Ensure that types are stable for EEReco, meaning that they are
EEjet{T} for a specific T, not just generic EEjet

graeme-a-stewart committed:
Fixes for type variation from CLI

Allow instrumented-jetreco to have a type argument that makes the particle
collection have that specific type, allowing for switches to different Floats

N.B, Currently Float32 is slower than Float64

graeme-a-stewart committed:
Remove Infiltrator temporary

graeme-a-stewart committed:
Fixes for pp reconstruction to use typed PseudoJets

It is very important to pass the fully typed PseudoJet{T} around, not a generic PseudoJet

graeme-a-stewart committed:
Define eltype() for PseudoJet and EEjet

Returns the parametrised element type

graeme-a-stewart committed:
Add more documentation for constructors

Also support type switching constructor for PseudoJet

graeme-a-stewart committed:
Formatting!

graeme-a-stewart committed:
Remove non-parametrised outer constructor

At the full constructor level the type parameter really should be known,
the use case for the non-explicit parametrised version is weak, so
it is removed. (This was also causing a docstring warning.)

graeme-a-stewart committed:
Make use of the known jet type parameter

The particle type is know in the function signature of the underlying reconstruction

graeme-a-stewart commented on src/EEAlgorithm.jl:
```diff
@@ -264,15 +265,18 @@ end
 
+    # Numerical type?
+    ParticleType = eltype(particles[1])
+
```
You are right - this can just be `T` here

graeme-a-stewart commented on src/Pseudojet.jl:
```diff
@@ -18,78 +18,119 @@ const _invalid_phi = -100.0
+
+# Returns
+A `PseudoJet` object.
+"""
```
The error is odd as it's quite possible to have these two different constructors. However, there's really no good case for the non-parametrised version, so I have removed it.

graeme-a-stewart committed:
Fix for jetreco.jl jet types

We explicitly use Float64 here, because this is deliberately the simple case.

gojakuch commented on examples/jetreco.jl:
```diff
@@ -140,17 +140,19 @@ function main()
+    # If we don't have an algorithm we default to PseudoJet
+    if !isnothing(args[:algorithm])
+        JetReconstruction.is_ee(args[:algorithm])
+        jet_type = EEjet{Float64}
```
just a question: why is it explicitly `Float64` here, shouldn't this be smth like `eval(args[:type])` for generality?

graeme-a-stewart replied:
The reason is that I wanted `jetreco.jl` to be the simplest and most explicit example possible. `instrumented-jetreco.jl` is the one with all the bells and whistles.

It might be useful to add a note though, that any subtype of `Real` is allowed.

gojakuch commented on examples/jetreco.jl:
```diff
@@ -140,17 +140,19 @@ function main()
+        jet_type = EEjet{Float64}
     else
-        jet_type = PseudoJet
+        jet_type = PseudoJet{Float64}
```
same question :)

graeme-a-stewart replied:
same answer 🤣

graeme-a-stewart committed:
Adapt tests to allow variable float types

Numerical type can be passed as a parameter to the test struct
This is passed down to read the events in the required precision.
Add some tests for e+e- and pp in Float32

graeme-a-stewart committed:
Fix type comparison in algorithm entry point

Had used wrong comparison - 'isa' instead of '<:'!

graeme-a-stewart committed:
Formatting

graeme-a-stewart committed:
Fix tests for parametrised types

graeme-a-stewart commented:
With the last set of fixes for correct types comparison, performance for the parameterised types are now the same as for the fixed `Float64`

Fixed:
```sh
 ~/tmp/JetReconstruction.jl/examples/ [main] ./benchmark.sh
pp 14TeV Tiled
Processed 100 events 16 times
Average time per event 180.95424500000001 ± 7.688918170282834 μs
Lowest time per event 171.03375 μs
pp 14 TeV Plain
Processed 100 events 16 times
Average time per event 664.906433125 ± 6.51083132764368 μs
Lowest time per event 658.1575 μs
ee H Durham
Processed 100 events 16 times
Average time per event 27.603828125 ± 0.21974330243013962 μs
Lowest time per event 27.315 μs
```

Parameterised:
```sh
 ~/.julia/dev/JetReconstruction/examples/ [general-numeric-types] ./benchmark.sh
pp 14TeV Tiled
Processed 100 events 16 times
Average time per event 177.1443225 ± 6.569998191079404 μs
Lowest time per event 170.38208999999998 μs
pp 14 TeV Plain
Processed 100 events 16 times
Average time per event 664.585339375 ± 4.866904941313732 μs
Lowest time per event 659.27709 μs
ee H Durham
Processed 100 events 16 times
Average time per event 27.670936875 ± 0.31133985727780145 μs
Lowest time per event 27.28125 μs
```
